### PR TITLE
fix: redirect page and first-hit normalization

### LINKED ISSUE #148: bug: iframe loading on localhost:3000
https://github.com/ipfs-shipyard/helia-service-worker-gateway/blob/92679927312c5f2e8363dc7838b50450e188b8a4/src/redirectPage.tsx#L14

testing http://localhost:3000/ipfs/bafkqablimvwgy3y results in redirect page trying to load config at url: `src="http://http://localhost:3000/ipfs/bafkqablimvwgy3y:3000#/ipfs-sw-config@origin=http%3A%2F%2Flocalhost%3A3000"`

replacing `${window.location.protocol}//${parentDomain}` with `${parentDomain}` correctly renders the iframe.. but then results in a bunch of nested iframes:

image

### LINKED ISSUE #150: bug: path requests should not try to render iframe
going to http://localhost:3000/ipfs/bafkqablimvwgy3y should render the redirect page if the service worker is not registered, but should **not** render the config iframe.

### PR DESCRIPTION
**Pull request by SgtPooki**
## Title

fix: redirect page and first-hit normalization

## Description

Fixes config page loading and normalizes first hits landing on redirectPage.

Fixes https://github.com/ipfs-shipyard/service-worker-gateway/issues/148
Fixes https://github.com/ipfs-shipyard/service-worker-gateway/issues/150

### Summary of changes

1. Adds new page and shouldRender() function for _redirects interstitial page.
2. Fixes config page loading on localhost.
3. Removes redirect logic from service-worker context page. This is handled by the redirectPage.

## Notes & open questions

This normalizes how redirects are handled. Now, the first hit on a redirect page will always land on the redirect page, and subsequent hits will be redirected to the target page. Previously, the first hit for ?helia-sw would actually load the helperUI page, which is not ideal.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if necessary (this includes comments as well)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: src/pages/redirect-page.tsx [hunk 1]
```diff
@@ -7,11 +7,18 @@ import { getSubdomainParts } from '../lib/get-subdomain-parts.js'
 import { isConfigPage } from '../lib/is-config-page.js'
 import { error, trace } from '../lib/logger.js'
 
-const ConfigIframe = (): JSX.Element => {
+const ConfigIframe = (): JSX.Element | null => {
```
Comment (SgtPooki) [inline src/pages/redirect-page.tsx]:
```suggestion
const ConfigIframe = (): JSX.Element => {
```

File: src/lib/routing-render-checks.ts [hunk 2]
```diff
@@ -11,7 +11,15 @@ export async function shouldRenderRedirectPage (): Promise<boolean> {
 
 export async function shouldRenderConfigPage (): Promise<boolean> {
   const { isConfigPage } = await import('../lib/is-config-page.js')
+  const url = new URL(window.location.href)
+  const heliaSw = url.searchParams.get('helia-sw')
 
   const isRequestToViewConfigPage = isConfigPage(window.location.hash)
-  return isRequestToViewConfigPage
+  return isRequestToViewConfigPage || heliaSw != null
```
Comment (SgtPooki) [inline src/lib/routing-render-checks.ts]:
```suggestion
  return isRequestToViewConfigPage
```

File: src/lib/routing-render-checks.ts [hunk 3]
```diff
@@ -11,7 +11,15 @@ export async function shouldRenderRedirectPage (): Promise<boolean> {
 
 export async function shouldRenderConfigPage (): Promise<boolean> {
   const { isConfigPage } = await import('../lib/is-config-page.js')
+  const url = new URL(window.location.href)
+  const heliaSw = url.searchParams.get('helia-sw')
```
Comment (SgtPooki) [inline src/lib/routing-render-checks.ts]:
```suggestion
```

File: src/pages/redirects-interstitial.tsx [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  const windowLocation = translateIpfsRedirectUrl(window.location.href)
+  if (windowLocation.href !== window.location.href) {
+  /**
+   * We're at a domain with ?helia-sw=, we can reload the page so the service worker will
+   * capture the request
+   */
+    window.location.replace(windowLocation.href)
+  }
+
+  return (<>First-hit on IPFS hosted service-worker-gateway. Reloading</>)
```
Comment (SgtPooki) [inline src/pages/redirects-interstitial.tsx:17]: we could probably add a spinner or something better here.

File: src/pages/redirects-interstitial.tsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ */
+function translateIpfsRedirectUrl (urlString: string): URL {
+  const url = new URL(urlString)
+  const heliaSw = url.searchParams.get('helia-sw')
+  if (heliaSw != null) {
+    url.searchParams.delete('helia-sw')
+    url.pathname = heliaSw
+  }
+  return url
+}
```
Comment (SgtPooki) [inline src/pages/redirects-interstitial.tsx:34]: this was moved from ipfs-hosted-redirect-utils.ts without modification

File: src/index.tsx [hunk 6]
```diff
@@ -13,11 +13,13 @@ const root = ReactDOMClient.createRoot(container)
 const LazyConfig = React.lazy(async () => import('./pages/config.jsx'))
 const LazyHelperUi = React.lazy(async () => import('./pages/helper-ui.jsx'))
 const LazyRedirectPage = React.lazy(async () => import('./pages/redirect-page.jsx'))
+const LazyInterstitial = React.lazy(async () => import('./pages/redirects-interstitial.jsx'))
 
 const routes: Route[] = [
   { default: true, component: LazyHelperUi },
-  { path: '#/ipfs-sw-config', shouldRender: async () => (await import('./lib/routing-render-checks')).shouldRenderConfigPage(), component: LazyConfig },
-  { shouldRender: async () => (await import('./lib/routing-render-checks')).shouldRenderRedirectPage(), component: LazyRedirectPage }
+  { shouldRender: async () => (await import('./lib/routing-render-checks.js')).shouldRenderRedirectsInterstitial(), component: LazyInterstitial },
```
Comment (SgtPooki) [inline src/index.tsx:24]: We could probably import this check function immediately since it's tiny but this is consistent at least.

**Review by SgtPooki — COMMENTED** [review]
self review

**Comment by SgtPooki** [comment]
I need this in to fix config page loading on subdomains on localhost... will merge shortly.

File: src/index.tsx (hunk 6 above)
Comment (2color) [reply src/index.tsx:24]: Sounds like a reasonable way to avoid an additional round trip.

**Comment by 2color** [comment]
Nice! From my testing this resolves the redirection loop.

**Pull request merged**